### ISSUE TITLE
Font.FromLogFont(object lf) cause exception in .Net Core

### ISSUE THREAD

**Issue by Lily721** [dotnet/winforms#2136]
See the LogFont struct below:		
```
[Serializable()]
[StructLayout(LayoutKind.Sequential, CharSet=CharSet.Auto)]
public struct LOGFONT 
{
	public int				lfHeight;
	public int				lfWidth;
	public int				lfEscapement;
	public int				lfOrientation;
	public int				lfWeight;
	public byte				lfItalic;
	public byte				lfUnderline;
	public byte				lfStrikeOut;
	public byte				lfCharSet;
	public byte				lfOutPrecision;
	public byte				lfClipPrecision;
	public byte				lfQuality;
	public byte				lfPitchAndFamily;
	[MarshalAs(UnmanagedType.ByValTStr, SizeConst=32)]
        public string			lfFaceName;

        /// <summary>
        ///   Add this methods in order to recudce the warning count of Fon32 class.
        /// </summary>
        public bool Equals(LOGFONT logFont)
        {
            //return ...
       }
}
```
If we use Font.FromLogFont() to create a font, then it will goes to an exception in .Net Core Windows Forms:
```System.ArgumentException: Object contains non-primitive or non-blittable data.  ```

I debugged .Net Core source code, and it seems like following code lead to this issue:
```GCHandle gch = GCHandle.Alloc (logFont, GCHandleType.Pinned);```

If I cannot use [MarshalAs(UnmanagedType.ByValTStr, SizeConst=32)] in my font struct how should I define my struct? Is this a bug?

**Comment by RussKie** [comment]
@Lily721 please fill in the bug report completely.
In particular we  need to know the runtime version you're having problems with.

**Comment by Lily721** [comment]
@RussKie , please see the stack tack below:
```
System.ArgumentException: Object contains non-primitive or non-blittable data.
   at System.Runtime.InteropServices.GCHandle..ctor(Object value, GCHandleType type)
   at System.Runtime.InteropServices.GCHandle.Alloc(Object value, GCHandleType type)
   at System.Drawing.Font.FromLogFont(Object lf, IntPtr hdc)
   at System.Drawing.Font.FromLogFont(Object lf)```

And I'm using .Net Core 3.0.1 at Win10 64-bit

**Comment by RussKie** [comment]
dotnet –info

**Comment by weltkante** [comment]
@RussKie looks like a bug in CoreFX. They are checking [IsValueType](https://github.com/dotnet/corefx/blob/dc7005cd64787880dfc27f19120944d46fb8f200/src/System.Drawing.Common/src/System/Drawing/Font.Windows.cs#L314) to determine which code path to use, but not all structs are blittable (in particular the `MarshalAs` attribute in the OP struct definition causes it to be non-blittable so the fast codepath will fail).

CoreFX needs to fix this by changing `IsValueType` to actually check if the struct is blittable.

@Lily721 if you don't use that type anywhere else you could, as a workaround, try defining it as a class instead of a struct to force CoreFX take the correct code path. (However if you do use it somewhere else it might not be safe to change to class, depends on how you use it.)

**Comment by Lily721** [comment]
@weltkante @RussKie , yes, we do use this type in somewhere else, and I don't think it is safe to change it to class, because we may change the property value of this type. But, anyway I still tried your workaround, it failed again, this time exception occurred in Application.cs at following code:
```
if (!component.FPreTranslateMessage(ref msg))
{
    UnsafeNativeMethods.TranslateMessage(ref msg);
    if (unicodeWindow)
    {
        UnsafeNativeMethods.DispatchMessageW(ref msg);
    }
    else
    {
        UnsafeNativeMethods.DispatchMessageA(ref msg);
    }
}
```
I'd like to know will you fix this issue in next release? or when it will be fixed?

**Comment by weltkante** [comment]
> it failed again, this time exception occurred in Application.cs at following code

This sounds like an entirely different problem and not related to Font logic, please create a separate issue with details about the problem - ideally with a way to reproduce, but at least a stack trace and under what circumstances the error occurs. I don't remember any known problem in the code block you posted, just saying "something fails there" isn't actionable.

> I'd like to know will you fix this issue in next release? or when it will be fixed?

The issue will probably be moved to the CoreFX repo (since the Font class is not part of WinForms), the new issue then should be linked from here so you can reach it and subscribe it. Once they decided which versions get a fix this usually is noted on the issue labels/milestones in the top right.

(Note that I'm not a member of either repo, I'm just helping out a bit)

**Comment by JeremyKuhne** [comment]
> I still tried your workaround, it failed again, this time exception occurred in Application.cs at following code:

How did it fail? Can you include the details?

You can keep it a struct if you used a fixed size array for `lfFaceName` to make it blittable. `System.Drawing` does this [here](https://github.com/dotnet/corefx/blob/07262aa6bce06f1e270e6a9b494bfe6eb1260b08/src/System.Drawing.Common/src/System/Drawing/Gdiplus.cs#L643).

If you need to read/write the face name you can add methods or a property that translates the fixed array to/from a string. I'm happy to help you write it correctly.

This API was never well documented- when I improved the performance for WinForms here I had to do a fair amount of guessing how people would call it. :) I'll look into making your struct work as is, but it will likely be quite a while before a fix is in a release build.

**Comment by RussKie** [comment]
@weltkante  @JeremyKuhne thank you

> I'll look into making your struct work as is, but it will likely be quite a while before a fix is in a release build.

@JeremyKuhne I think this bug will meet the bar for 3.1, so if we have a fix in time, it go out in 3.1.

**Comment by JeremyKuhne** [comment]
> @JeremyKuhne I think this bug will meet the bar for 3.1, so if we have a fix in time, it go out in 3.1.

I don't think it will meet the bar for Core (this is a problem in System.Drawing), but I'll see if I can come up with fix nonetheless.

**Comment by JeremyKuhne** [comment]
The fix looks to be pretty simple (linked), so it might meet the bar. I'll definitely submit it for a bar check. (We don't have a lot of time before 3.1 completes, hence the hesitance to commit to when it goes in.)

**Comment by weltkante** [comment]
Just for the record, I think this bug breaks this API, so fixing it in 3.1 is a good thing. For what its worth we would hit this bug as well if we were to port our script editor control without rewriting the text rendering interop (currently targeting to 5.0 though so not an immediate problem for myself)

> This API was never well documented- when I improved the performance for WinForms here I had to do a fair amount of guessing how people would call it.

Using a non-blittable struct was the recommended way to do interop on Desktop Framework, using a class or a blittable struct would be the exceptional case. When porting an application you wouldn't start by rewriting your whole interop code using blittable structs.

**Comment by RussKie** [comment]
/cc: @merriemcgaw you may need to bring this to the Tactics' attention

**Comment by danmosemsft** [comment]
Reopening pending the question to @merriemcgaw : do you want us to port this to 3.1?

If so @JeremyKuhne can make a PR, but it will need to go in before 4pm tomorrow.

**Comment by merriemcgaw** [comment]
@danmosemsft  Yes, I really do!

**Comment by JeremyKuhne** [comment]
This was approved for 3.1. https://github.com/dotnet/corefx/pull/42007 is the 3.1 PR. The fix is in for master (5.0).
